src/bucket/BucketController.ts

```diff
@@ -0,0 +1,254 @@
+import {AmazonS3Backend} from './../common/AmazonS3Backend';
+
+export class BucketController {
+  /**
+   * Dependencies we want passed to the constructor.
+   * @see http://docs.angularjs.org/guide/di
+   */
+  public static $inject = [
+    '$rootScope'
+  ];
+
+  /**
+   * Backend name. Passed in as component binding.
+   */
+  public backendName: string;
+
+  /**
+   * Bucket name. Passed in as a component binding.
+   */
+  public bucketName: string;
+
+  /**
+   * AWS region. Passed in as a component binding.
+   */
+  public awsRegion: string;
+
+  /**
+   * AWS Access Key ID. Passed in as a component binding.
+   */
+  public awsAccessKeyId: string;
+
+  /**
+   * AWS Secret Access Key. Passed in as a component binding.
+   */
+  public awsSecretAccessKey: string;
+
+  /**
+   * Flag used to indicate a background operation is running.
+   */
+  public working: boolean;
+
+  /**
+   * Error encountered running background operation.
+   */
+  public error: Error;
+
+  /**
+   * Bucket.
+   */
+  public bucket: Bucket;
+
+  /**
+   * Current working folder.
+   */
+  public currentFolder: Folder;
+
+  /**
+   * Folder objects in the current working path.
+   */
+  public folders: Folder[];
+
+  /**
+   * File objects in the current working path.
+   */
+  public files: File[];
+
+  /**
+   * Settings for uploading files.
+   */
+  public uploadConfig: IUploadConfig;
+
+  /**
+   * Used to specify the name of new folders.
+   */
+  public folderName: string;
+
+  /**
+   * Backend.
+   */
+  private backend: IBackend;
+
+  /**
+   * Create an instance of the controller.
+   */
+  constructor(private $rootScope: ng.IScope) {
+    this.working = false;
+    this.error = null;
+    this.bucket = null;
+    this.currentFolder = new Folder(new Path('/'));
+    this.folders = [];
+    this.files = [];
+    this.uploadConfig = null;
+    this.folderName = '';
+  }
+
+  /**
+   * Initialize the controller.
+   */
+  $onInit() {
+    // create the backend
+    if (this.backendName === 's3') {
+      this.backend = new AmazonS3Backend(
+        this.awsRegion,
+        this.awsAccessKeyId,
+        this.awsSecretAccessKey);
+    } else {
+      throw `Unknown backend: ${this.backendName}`;
+    }
+
+    // initial load
+    this.loadContents();
+  }
+
+  /**
+   * Load bucket and objects at working path.
+   */
+  public loadContents(): Promise<any> {
+    this.working = true;
+    this.error = null;
+    return this.backend.getBucket(this.bucketName)
+      .then((bucket: Bucket) => {
+        // store bucket
+        this.bucket = bucket;
+
+        // retrieve upload settings
+        this.uploadConfig = this.backend.getUploadConfig(
+          this.bucket,
+          this.currentFolder);
+
+        // load current folder contents
+        return this.backend.getContents(bucket, this.currentFolder);
+      })
+      .then((contents: IFolderContents) => {
+        function compareObjectNames (a: IBucketObject, b: IBucketObject) {
```

Comment:
This is what I was referring to in the outdated commit about refactoring this function out instead of instantiating it here every time this promise runs. Doubt that it's a hit on performance. I only bring this up because you use a similar pattern in the File module but it's declared outside the promise.